[sqladmin-912] - Add perform context functions, DateRange filter, JSON export,  and other tweaks...

sergei-vasilev-dev opened the pull request:
https://github.com/aminalaee/sqladmin/issues/912


I think better to perform template context on the model level. Please, take a look and say what you think. 
Thanks!

sergei-vasilev-dev committed:
[sqladmin-912] - Add perform context functions to the Modelview class.

aminalaee commented:
Hey @sergey-vasiliev-dualboot ,
I think this is nice, Can you please update the docs and explain the example use-case there?

sergei-vasilev-dev committed:
feat: Advanced filtering system with DateRange, ManyToMany, RelatedModel filters

- Add UniqueValuesFilter with Integer/Float support and formatting
- Add ManyToManyFilter for filtering through junction tables
- Add RelatedModelFilter for filtering by related model columns
- Add DateRangeFilter for date/datetime range filtering
- Enhance ForeignKeyFilter with multiple value selection
- Add _safe_join() to prevent duplicate JOINs
- Add async_search_query() for custom async search
- Enhance pretty_export with JSON support
- Add comprehensive documentation with cookbook examples

sergei-vasilev-dev committed:
test: Add 18 tests for new filtering features and clean up docstrings

- Add tests for UniqueValuesFilter with Integer/Float
- Add tests for enhanced ForeignKeyFilter
- Add tests for DateRangeFilter
- Add tests for _safe_join() preventing duplicate JOINs
- Add tests for add_relation_loads() optimization
- Add tests for async_search_query()
- Add test for pretty_export_json()
- Clean up verbose docstrings in favor of self-documenting code
- Move imports from functions to module level

All 248 tests passing.

sergey-vasiliev-dualboot commented:
# Enhanced Filtering System and ModelView Improvements

## Summary

This PR introduces significant enhancements to SQLAdmin's filtering capabilities and query optimization. All changes are **fully backward compatible** and extend existing functionality without breaking changes.

## 🎯 New Features

### Advanced Filter Types

#### 1. **UniqueValuesFilter** (Enhanced)
Improved version with support for numeric types and custom formatting:
- ✅ Integer and Float column support
- ✅ Custom value display with `lookups_ui_method`
- ✅ Float rounding with `float_round_method`
- ✅ Custom ordering with `lookups_order`
- ✅ Multiple value selection

```python
from sqladmin.filters import UniqueValuesFilter
import math

UniqueValuesFilter(
    Product.price,
    lookups_ui_method=lambda v: f"${v:.2f}",
    float_round_method=lambda v: math.floor(v),
    lookups_order=Product.price
)
```

#### 2. **ManyToManyFilter** (NEW)
Filter through many-to-many relationships using junction tables:

```python
from sqladmin.filters import ManyToManyFilter

ManyToManyFilter(
    column=User.id,
    link_model=UserRole,
    local_field="user_id",
    foreign_field="role_id",
    foreign_model=Role,
    foreign_display_field=Role.name
)
```

#### 3. **RelatedModelFilter** (NEW)
Filter by columns in related models through automatic JOIN:

```python
from sqladmin.filters import RelatedModelFilter

RelatedModelFilter(
    column=User.address,
    foreign_column=Address.city,
    foreign_model=Address
)
```

#### 4. **DateRangeFilter** (NEW)
Filter by date/datetime ranges with start and end values:

```python
from sqladmin.filters import DateRangeFilter

DateRangeFilter(
    Order.created_at,
    title="Order Date"
)
```

#### 5. **Enhanced ForeignKeyFilter**
Now supports multiple value selection and custom ordering:

```python
ForeignKeyFilter(
    Product.category_id,
    Category.name,
    foreign_model=Category,
    lookups_order=Category.name  # NEW: Sort alphabetically
)
```

### ModelView Query Optimization

#### **_safe_join()**
Prevents duplicate JOINs in complex queries with related models:
- Automatically used in `search_query()` and `sort_query()`
- Significantly improves query performance
- No configuration needed - works automatically

#### **add_relation_loads()**
Optimizes relationship loading using `selectinload`:
- Reduces N+1 query problems
- Automatically applied in `list()` and `get_model_objects()`
- Works transparently

#### **async_search_query()**
Optional async search for custom implementations:

```python
class UserAdmin(ModelView, model=User):
    async_search = True
    
    async def async_search_query(self, stmt, term, request):
        # Custom async search logic
        return stmt.filter(...)
```

### Pretty Export Enhancements

#### **JSON Export with Formatting**
Pretty export now supports JSON format:
- Applies column labels as keys
- Uses custom formatters
- Maintains all formatting logic

```python
class OrderAdmin(ModelView, model=Order):
    use_pretty_export = True
    export_types = ["csv", "json"]
```

#### **Better Relation Handling**
- Support for `related_model_relations` attribute
- Improved list handling in related fields

## 📊 Testing

### Test Coverage
- ✅ **248 tests passing** (added 18 new tests)
- ✅ All existing tests still pass
- ✅ No regression issues

### New Tests Added
- `test_unique_values_filter_*` (6 tests)
- `test_many_to_many_filter_*` (2 tests)
- `test_related_model_filter_*` (1 test)
- `test_date_range_filter_*` (2 tests)
- `test_safe_join_*` (2 tests)
- `test_add_relation_loads` (1 test)
- `test_async_search_query_*` (1 test)
- `test_*_uses_safe_join` (2 tests)
- `test_pretty_export_json_*` (1 test)

## 📚 Documentation

### New Documentation
- ✅ `docs/cookbook/advanced_filtering.md` - Comprehensive filtering guide with examples
- ✅ `docs/cookbook/readonly_views.md` - Read-only view patterns and best practices

### Updated Documentation
- ✅ `docs/configurations.md` - Added "Advanced Filters" and "Async Search" sections
- ✅ `docs/working_with_templates.md` - Examples for `perform_*_context` methods
- ✅ `README.md` - Mentioned advanced filtering capabilities

## 🔧 Code Quality

- ✅ Linter checks pass (`ruff`)
- ✅ Type checking passes (`mypy`)
- ✅ Code formatted per project standards
- ✅ All docstrings concise and meaningful
- ✅ No imports inside functions

## ⚠️ Breaking Changes

**None!** All changes are fully backward compatible:
- New methods are optional
- New attributes have sensible defaults
- Existing API unchanged
- All existing code works without modification

## 🔄 Migration Guide

No migration needed! Simply start using new features:

```python
from sqladmin import ModelView
from sqladmin.filters import UniqueValuesFilter, ManyToManyFilter

class MyAdmin(ModelView, model=MyModel):
    column_filters = [
        UniqueValuesFilter(MyModel.price),  # Just use it!
        ManyToManyFilter(...)  # Works alongside existing filters
    ]
```

## 📈 Performance Impact

### Improvements
- ✅ **Safe JOIN** - Prevents duplicate JOINs that slow down queries
- ✅ **Relation Loading** - Eliminates N+1 query problems
- ✅ **Zero overhead** - New features are opt-in only

### No Negative Impact
- Default behavior unchanged
- Existing queries maintain same performance
- No additional dependencies

## 🎯 Use Cases

### E-commerce Platform
```python
class ProductAdmin(ModelView, model=Product):
    column_filters = [
        UniqueValuesFilter(Product.price, lookups_ui_method=lambda v: f"${v:.2f}"),
        ManyToManyFilter(...),  # Filter by tags
        RelatedModelFilter(...),  # Filter by supplier country
        DateRangeFilter(Product.created_at)
    ]
```

### Analytics Dashboard
```python
class OrderAdmin(ModelView, model=Order):
    async_search = True
    use_pretty_export = True
    export_types = ["csv", "json"]
```

## 📝 Implementation Details

### Files Modified
- `sqladmin/filters.py` (+265 lines)
- `sqladmin/models.py` (+82 lines)
- `sqladmin/pretty_export.py` (+38 lines)
- `sqladmin/application.py` (+8 lines)
- `tests/` (+272 lines)

## ✅ Checklist

- [x] Code follows project style guidelines
- [x] All tests pass (248 passing)
- [x] Linter checks pass
- [x] Type checking passes
- [x] Documentation updated
- [x] Backward compatibility maintained
- [x] No breaking changes
- [x] Examples provided in docs
- [x] Self-documenting code with minimal docstrings

## 🚀 Ready for Review

This PR has been thoroughly tested and documented. All changes maintain backward compatibility while significantly extending SQLAdmin's filtering and query optimization capabilities.

---

**Looking forward to feedback!**

sergei-vasilev-dev committed:
test: Improve coverage to 99% with comprehensive filter tests

- Add 40+ new tests for filter edge cases
- Add tests for all utility functions
- Add tests for DateRangeFilter with various inputs
- Add tests for RelatedModelFilter Boolean handling
- Add tests for pretty_export_json with formatters
- Add pragma: no cover for impossible-to-test cases

All 291 tests passing with 99% coverage.

rusanpas commented:
This seems like a great improvement, but I have a question.

Shouldn't the filters be accompanied by changes to the templates?

The date filter, without modifying the template to add, for example, a Flatpickr (range), would leave the filter incomplete unless a custom template is created, right?

sergei-vasilev-dev committed:
feat: Add DateRangeFilter UI, Table support, and demo application

Core Features:
- Add DateRangeFilter with datetime-local inputs UI
- Add hasattr to Jinja2 globals for cleaner templates
- Add SQLAlchemy Table object support in filters
- Handle DateRangeFilter params (_start, _end) in list()
- Fix ManyToManyFilter to work with Table objects

Demo Application:
- Add comprehensive demo_app/ with all features
- 8 models (User, Product, Order, etc) with FK and M2M
- 9 admin views showcasing all filter types
- Sample data for immediate testing
- Quick start guide and documentation

Testing:
- Add tests for Table object handling
- Add tests for DateRangeFilter integration
- All 292 tests passing, 99% coverage

DateRangeFilter and ManyToManyFilter now work with association tables!

To test: cd demo_app && pip install -e .. fastapi uvicorn sqlalchemy && python main.py
Then open: http://localhost:8000/admin

mmzeynalli commented on sqladmin/filters.py:
```diff
@@ -187,20 +208,323 @@ async def lookups(
         ]
 
     async def get_filtered_query(self, query: Select, value: Any, model: Any) -> Select:
+        if value == "" or value == [""] or not value:
```
`value == ""` is covered by `not value`. So, change pls:

`if not value or value == [""]`

mmzeynalli commented on sqladmin/templates/sqladmin/list.html:
```diff
@@ -262,8 +262,39 @@ <h3 class="card-title">Filters</h3>
                   </form>
                 </div>
               </div>
+              {% elif hasattr(filter, 'is_date_filter') and filter.is_date_filter %}
```
dont now if getattr works here

mmzeynalli commented on sqladmin/filters.py:
```diff
@@ -187,20 +208,323 @@ async def lookups(
+        return joined_query.filter(filter_condition)
+
+
+class DateRangeFilter:
```
Just curious: Do you think current design would support RangeFilter to other types as well? Let's say, I want all users between age 20 and 30. Or filter range of ForeignKey values (20 < User.profile.age < 30).

mmzeynalli committed:
Merged main

mmzeynalli committed:
Merge branch 'main' into feature/sqladmin-912_perform_template_context
